File: Tree/Leetcode-297-SerializeAndDeserializeBinaryTree/Leetcode-297-SerializeAndDeserializeBinaryTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Codec {
public:
    void serial(TreeNode* node, string& str){
        if(node == nullptr){
            str += "null";
        }else{
            str += to_string(node->val) + ",";
            serial(node->left, str); str += ",";
            serial(node->right, str);
        }
    }
    
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string str = "";
        serial(root, str);
        return str;
    }

    std::vector<std::string> string_split(std::string str, std::string delimiter){
        size_t pos = 0;
        std::string token;
        std::vector<std::string> result;

        while(true){
            pos = str.find(delimiter);
            token = str.substr(0, pos);
            result.push_back(token);
            if(pos == string::npos) break;
            str.erase(0, pos+delimiter.length());
        }
        return result;
    }
    
    TreeNode* deserial(queue<string>& q){
        string token = q.front(); q.pop();
        if(token == "null") return nullptr;
        TreeNode* root = new TreeNode(stoi(token));
        root->left = deserial(q);
        root->right = deserial(q);
        return root;
    };
    
    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        vector<string> tokens = string_split(data, ",");
        queue<string, deque<string>> q(deque<string>(tokens.begin(), tokens.end()));
        return deserial(q);
    }
};

File: 2월4주차/leetcode-297-Serialize and Deserialize Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

class Codec: # 처음 코드. dfs
    def serialize(self, root):
        myarray = []
        def dfs(node):
            if not node:
                myarray.append("null")
            else:
                myarray.append(node.val)
                dfs(node.left)
                dfs(node.right)
        dfs(root)
        return ','.join(map(str, myarray))

    def deserialize(self, data):
        darray = data.split(',')
        darray = [int(element) if element != "null" else None for element in darray]

        def dfs(node):
            if len(darray) == 0:
                return
            
            val = darray.pop(0)
            if val != None:
                node.left = TreeNode(val)
                dfs(node.left)

            val = darray.pop(0)
            if val != None:
                node.right = TreeNode(val)
                dfs(node.right)

        val = darray.pop(0)
        if val != None:
            root = TreeNode(val)
            dfs(root)
            return root

class Codec: # bfs로도 한번 짜봄
    def serialize(self, root):
        dq = collections.deque([root])
        serArray = []
        while dq:
            node = dq.popleft()
            if node != None:
                serArray.append(node.val)
                dq.append(node.left)
                dq.append(node.right)
            else:
                serArray.append("None")
        return ','.join(map(str, serArray))

    def deserialize(self, data):
        deserArray = data.split(',')
        deserArray = [int(element) if element != "None" else None for element in deserArray]
        deserArray = collections.deque(deserArray)
        if deserArray[0] == None:
            return None

        root = TreeNode(deserArray.popleft())
        nodedq = collections.deque([root])
        while nodedq:
            node = nodedq.popleft()

            val = deserArray.popleft()
            if val != None:
                node.left = TreeNode(val)
                nodedq.append(node.left)
            
            val = deserArray.popleft()
            if val != None:
                node.right = TreeNode(val)
                nodedq.append(node.right)

        return root